File: apps/mobile/ios/Runner/Native/Camera/ReplayBufferHostApiImpl.swift
import Flutter
import Foundation

final class ReplayBufferHostApiImpl: NSObject, ReplayBufferHostApi, @unchecked Sendable {
  private let manager: CameraManager
  private let flutterApi: ReplayBufferFlutterApi

  init(manager: CameraManager, messenger: FlutterBinaryMessenger) {
    self.manager = manager
    self.flutterApi = ReplayBufferFlutterApi(binaryMessenger: messenger)
    super.init()
    manager.onReplaySaved = { [weak self] url, durationMs in
      DispatchQueue.main.async {
        self?.flutterApi.onReplaySaved(path: url.path, durationMs: Int64(durationMs)) { _ in }
      }
    }
    manager.onReplayFailed = { [weak self] error in
      DispatchQueue.main.async {
        self?.flutterApi.onReplayFailed(code: error.code, message: error.message) { _ in }
      }
    }
  }

  func enableReplayBuffer(seconds: Int64) throws {
    manager.setReplayWindow(seconds: Int(seconds))
  }

  func disableReplayBuffer() throws {
    manager.setReplayWindow(seconds: 0)
  }

  func saveReplay() throws {
    do {
      try manager.saveReplay()
    } catch let error as CameraNativeError {
      throw ReplayBufferPigeonError(code: "\(error.code)", message: error.message, details: nil)
    } catch {
      throw ReplayBufferPigeonError(code: "sessionFailed", message: error.localizedDescription, details: nil)
    }
  }
}

File: apps/mobile/android/app/src/main/kotlin/com/rarocamera/raro_mobile/replay/ReplayBufferHostApiImpl.kt
package com.rarocamera.raro_mobile.replay

import android.os.Handler
import android.os.Looper
import com.rarocamera.raro_mobile.camera.CameraManager
import com.rarocamera.raro_mobile.camera.CameraNativeException
import com.rarocamera.raro_mobile.camera.symbolicCode
import com.rarocamera.raro_mobile.generated.replay_buffer.FlutterError
import com.rarocamera.raro_mobile.generated.replay_buffer.ReplayBufferFlutterApi
import com.rarocamera.raro_mobile.generated.replay_buffer.ReplayBufferHostApi
import io.flutter.plugin.common.BinaryMessenger

class ReplayBufferHostApiImpl(
  private val cameraManager: CameraManager,
  messenger: BinaryMessenger,
) : ReplayBufferHostApi {
  private val main = Handler(Looper.getMainLooper())
  private val flutterApi = ReplayBufferFlutterApi(messenger)

  init {
    cameraManager.onReplaySaved = { path, durationMs ->
      main.post { flutterApi.onReplaySaved(path, durationMs) {} }
    }
    cameraManager.onReplayFailed = { code, message ->
      main.post { flutterApi.onReplayFailed(code, message) {} }
    }
  }

  override fun enableReplayBuffer(seconds: Long) {
    cameraManager.enableReplayBuffer(seconds.toInt())
  }

  override fun disableReplayBuffer() {
    cameraManager.disableReplayBuffer()
  }

  override fun saveReplay() {
    try {
      cameraManager.saveReplay()
    } catch (e: CameraNativeException) {
      throw FlutterError(code = e.symbolicCode(), message = e.message, details = null)
    }
  }
}
